applications/tari_merge_mining_proxy/src/proxy.rs

```diff
@@ -791,3 +807,45 @@ pub(super) async fn read_body_until_end(body: &mut Body) -> Result<BytesMut, MmP
     }
     Ok(bytes)
 }
+
+fn json_rpc_success_response(req_id: Option<i64>, result: json::Value) -> json::Value {
+    json!({
+       "id": req_id.unwrap_or(-1),
+       "jsonrpc": "2.0",
+       "result": result,
+       "status": "OK",
+       "untrusted": false
+    })
+}
+
+fn json_rpc_standard_error_response(
+    req_id: Option<i64>,
+    err: jsonrpc::error::StandardError,
+    data: Option<json::Value>,
+) -> json::Value
+{
+    let err = jsonrpc::error::standard_error(err, data);
+    json!({
+        "id":  req_id.unwrap_or(-1),
+        "jsonrpc": "2.0",
+        "error": err,
+    })
+}
+
+fn json_rpc_error_response(
+    req_id: Option<i64>,
+    err_code: i32,
+    err_message: &str,
+    err_data: Option<json::Value>,
+) -> json::Value
+{
+    json!({
+        "id":  req_id.unwrap_or(-1),
+        "jsonrpc": "2.0",
+        "error": {
+            "code": err_code,
+            "message": err_message,
+            "data": err_data,
```

Comment:
See previous comment.

Reply:
This is a standard JSON RPC error https://www.jsonrpc.org/specification#error_object

Reply:
True, though Monero does have errors mapped to codes and messages depending on what is wrong:
```
CORE_RPC_ERROR_CODE_WRONG_BLOCKBLOB
CORE_RPC_ERROR_CODE_WRONG_PARAM
CORE_RPC_ERROR_CODE_TOO_BIG_HEIGHT
CORE_RPC_ERROR_CODE_TOO_BIG_RESERVE_SIZE
CORE_RPC_ERROR_CODE_WRONG_WALLET_ADDRESS
CORE_RPC_ERROR_CODE_MINING_TO_SUBADDRESS
CORE_RPC_ERROR_CODE_INTERNAL_ERROR
CORE_RPC_ERROR_CODE_BLOCK_NOT_ACCEPTED
```

The definitions can be found in:
```
core_rpc_server_error_codes.h
```

And usage in:
```
core_rpc_server.cpp
```

Assuming configuration is correct, we will only get this error on failure:
```
CORE_RPC_ERROR_CODE_BLOCK_NOT_ACCEPTED
```

I'm not against using the above however, perhaps you want to underscore data?

Reply:
These helpers allow you to create any JSON RPC success or error response object. They have nothing to do with monerod or the submit_block call. I've added comments to these functions.

The submit_block call uses this function to create the "block not accepted" error response object [see here](https://github.com/tari-project/tari/pull/2616/files#diff-c12374a32500513f6277a01e907665e87a10f680738b198f60bab0213f4e1294R404)

The `data` field is part of the [JSON RPC spec](https://www.jsonrpc.org/specification#error_object), so it can stay.